xds/src/main/java/io/grpc/xds/XdsRoutingLoadBalancerProvider.java

```diff
@@ -0,0 +1,370 @@
+     * action of each of which is provided by the given map of actions.
+     */
+    @VisibleForTesting
+    XdsRoutingConfig(List<Route> routes,
+        Map<String, ChildConfig> actions) {
+      this.routes = ImmutableList.copyOf(routes);
+      this.actions = ImmutableMap.copyOf(actions);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      XdsRoutingConfig that = (XdsRoutingConfig) o;
+      return Objects.equals(routes, that.routes)
+          && Objects.equals(actions, that.actions);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(routes, actions);
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("routes", routes)
+          .add("actions", actions)
+          .toString();
+    }
+  }
+
+  static final class Route {
+
+    final String actionName;
+    final MethodName methodName;
+
+    @VisibleForTesting
+    Route(String actionName, MethodName methodName) {
+      this.actionName = actionName;
+      this.methodName = methodName;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      Route route = (Route) o;
+      return Objects.equals(actionName, route.actionName)
+          && Objects.equals(methodName, route.methodName);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(actionName, methodName);
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("actionName", actionName)
+          .add("methodName", methodName)
+          .toString();
+    }
+  }
+
+  static final class MethodName {
+
+    final String service;
+    final String method;
+
+    @VisibleForTesting
+    MethodName(String service, String method) {
+      this.service = service;
+      this.method = method;
+    }
+
+    boolean isDefault() {
+      return service.isEmpty() && method.isEmpty();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      MethodName that = (MethodName) o;
+      return Objects.equals(service, that.service)
+          && Objects.equals(method, that.method);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(service, method);
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("service", service)
+          .add("method", method)
+          .toString();
+    }
+  }
+
+  static final class ChildConfig {
```

Comment:
I think we should define a `ParsedConfig` class with `policyName` and `config` fields, similar to [`LbConfig`](https://github.com/grpc/grpc-java/blob/e29561fbca3eb7a536b38d0e0b91c95caef94700/core/src/main/java/io/grpc/internal/ServiceConfigUtil.java#L321). This could be useful for all hierarchical LB configs.

Reply:
Yes, that is also my plan, currently we have `AutoConfiguredLoadBalancerFactory.PolicySelection` which basically is the `ParsedConfig`. It has a deprecated field `rawConfig` just for the backward compatibility in legacy usecase.

Reply:
Also, defining types with name `ChildXXX` doesn't sound good, unless it's an internal implementation (private) or locally used. "Child" is meaningful only in the context of one policy's internal implementation.